### DISCUSSION TITLE
Integrate tree-editor into GLSP example

### DISCUSSION THREAD

**Discussion by julen9** [eclipse-glsp/glsp#1052] [Q&A] [answered]
Hi, I have `"java-emf-theia"` example running with custom metamodel, all good until here. My problem is I want to integrate `"tree-editor"` extension to my "java-emf-theia" example. What have I done:

Run isolated example of `"tree-editor"` via [https://github.com/eclipse-emfcloud/theia-tree-editor](https://github.com/eclipse-emfcloud/theia-tree-editor)

Get the folder `"tree-editor-example"` (I supouse this is the extension folder, with all necessary components to import the extension to other Theia IDE)

Add  `"tree-editor-example"` folder into my `"java-emf-theia"` example folder, at the same level of `"tasklist-theia"` folder (I suppouse this is the folder level to add the extensions)

![image](https://github.com/eclipse-glsp/glsp/assets/3693559/16fa718d-aca1-4d6d-8229-995918a4a2d5)

add in `"glsp-client/package.json"` new element to workspaces array, adding `"tree-editor-example"`

![image](https://github.com/eclipse-glsp/glsp/assets/3693559/732dd2be-1309-4e1e-a2d3-f804c5f1e787)

add in `"glsp-client/eda-browser-app/package.json"` new dependency `"tree-editor-example": "0.0.0"`

![image](https://github.com/eclipse-glsp/glsp/assets/3693559/5d6205bc-38c5-4a17-9c16-0278f80ebe3d)

after that, I launch `"yarn"` command to install dependencies and build it, later I run `"yarn start"` into the `"glsp-client"` folder. The server starts and I can see a new button in top menu but all buttons are disabled, I can't click it, and when I try to open my` .tree` file, it doesn't open the tree view, only can see the json code. In console I can see only this error:

> 2023-07-12T11:22:13.220Z root ERROR Error: No matching bindings found for serviceIdentifier: WorkspaceService
>     at _validateActiveBindingCount (http://localhost:3000/bundle.js:86584:23)
>     at _getActiveBindings (http://localhost:3000/bundle.js:86570:5)
>     at _createSubRequests (http://localhost:3000/bundle.js:86613:26)
>     at http://localhost:3000/bundle.js:86637:17
>     at Array.forEach (<anonymous>)
>     at http://localhost:3000/bundle.js:86636:26
>     at Array.forEach (<anonymous>)
>     at _createSubRequests (http://localhost:3000/bundle.js:86616:20)
>     at Object.plan (http://localhost:3000/bundle.js:86658:9)
>     at http://localhost:3000/bundle.js:86180:37

Please, what is the correct way to add `tree-editor` to my `java-emf-theia example`?
Thanks in advance!

**Comment by planger** [comment, marked as answer]
Thanks for detailing the steps you did! They make perfect sense and should be correct.

From the error I suspect that you pulled in multiple versions of Theia, in particular `@theia/workspace`. If you have multiple versions, the dependency injection breaks. Can you check in your `yarn.lock` whether everything resolved to exactly one `@theia/workspace` version and all other `@theia/*` extensions resolved to the same version?
I see that you've hard-pinned to `1.34.0` so all `@theia/*` extensions must be resolved to that version.

You can also run `yarn theia check:theia-version` to check whether there are multiple versions of Theia mixed in your project.

If you have multiple versions there, you'd need to find out why (e.g. using `yarn why  @theia/workspace`) and change the dependency. Alternatively you can use yarn resolutions to enforce a single version. See also https://github.com/eclipse-glsp/glsp-theia-integration#theia-version-compatibility

There was a similar discussion a few months ago: https://github.com/eclipse-glsp/glsp/discussions/816

Hope this helps!

**Reply by julen9** [reply]
yes, that works for me!! thanks a lot!
